### PR TITLE
add hooks for gradient and tendency filters

### PR DESCRIPTION
**Pull request by simonbyrne**
### Description

This should make it easier to do tendency-level filtering to emulate overintegration.

cc: @jkozdon @Zhengyu-Huang 

- [ ] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [ ] Unit tests are included OR N/A.
- [ ] Code is exercised in an integration test OR N/A.
- [ ] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Review by Zhengyu-Huang — APPROVED** [review]
That looks great to me!  Thank you for the effort. 
I will try this!

File: src/Numerics/DGMethods/DGModel.jl
Comment (jkozdon) [inline src/Numerics/DGMethods/DGModel.jl:316]: It might be good to modify `Filters.apply!` take an optional event so that this doesn't slow everything down.

The logic gets a tad complicated since if the user doesn't pass an `event` you would wait in the `apply!` otherwise you would return the new event.

**Comment by simonbyrne** [comment]
TODO:
- add `--cutoff-degree` option to the driver
- make filter asynchronous

File: src/Numerics/DGMethods/DGModel.jl
Comment (simonbyrne) [reply src/Numerics/DGMethods/DGModel.jl:316]: Probably better to have a separate `async_apply!` function

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
Okay, this is ready for review. It does need a test: I don't know what would be best to try out though.

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
@thomasgibson did you want to modify the isothermal zonal flow test?

**Comment by simonbyrne** [comment]
bors try

File: src/Driver/driver_configs.jl
Comment (valeriabarra) [inline src/Driver/driver_configs.jl:262]: Do we want a logical OR here or a logical AND? I thought both conditions were needed. Can we have a case where only one of the two is satisfied and we still can apply the filter?

File: src/Driver/driver_configs.jl
Comment (valeriabarra) [inline src/Driver/driver_configs.jl:279]: for consistency with the other prints, I'd use something like:
```suggestion
    horiz cutoff order        = %d
    vert cutoff order          = %d
```

File: src/Driver/driver_configs.jl
Comment (valeriabarra) [reply src/Driver/driver_configs.jl:262]: The same question applies to the following configurations

Comment (simonbyrne) [reply src/Driver/driver_configs.jl:262]: we only need to apply the filter if the cutoff order is < polyorder

Comment (valeriabarra) [reply src/Driver/driver_configs.jl:262]: Yes, but for both its components, horizontal and vertical, no?

**Comment by valeriabarra** [comment]
Note:

While testing this PR, I realized that in PR #1946 I did not check for backward compatibility, and unfortunately missed testing the `--degree` option that was already there. I'm pushing a fix to your branch right now.

File: src/Driver/driver_configs.jl
Comment (valeriabarra) [inline src/Driver/driver_configs.jl:203]: I'm testing this by running a different experiment than the `experiments/TestCase/isothermal_zonal_flow.jl` and this would break, if the default test arguments were used. I think we should give `Ncutoff` another default value, not the same as `N`.

File: src/Driver/driver_configs.jl
Comment (valeriabarra) [inline src/Driver/driver_configs.jl:217]: I'm testing this by running a different experiment than the `experiments/TestCase/isothermal_zonal_flow.jl` and this would break, if the default test arguments were used. I think we should give `Ncutoff` another default value, not the same as `N`.

**Review by valeriabarra — CHANGES_REQUESTED** [review]

File: src/Driver/driver_configs.jl
Comment (valeriabarra) [inline src/Driver/driver_configs.jl:282]: If we fall in this `else` condition, we should change the values of `cutofforder_horz` and `cutofforder_vert` to `nothing`, so that in the print that follows, the user indeed sees that in that case the filter was not applied. Otherwise, the way it is right now, the user would see the attempted cutoff filter degrees, not the actual (nothing) one and it would be misleading.

**Review by valeriabarra — DISMISSED** [review]

File: src/Driver/driver_configs.jl
Comment (valeriabarra) [reply src/Driver/driver_configs.jl:282]: I changed it in [c575dc3](https://github.com/CliMA/ClimateMachine.jl/pull/1953/commits/c575dc3358e3e202eea2e057078fda5b0550c311)

File: src/Driver/driver_configs.jl
Comment (simonbyrne) [reply src/Driver/driver_configs.jl:217]: we could make it `get_polyorder(N)`?

File: src/Driver/driver_configs.jl
Comment (simonbyrne) [reply src/Driver/driver_configs.jl:282]: Thanks!

File: src/Driver/driver_configs.jl
Comment (simonbyrne) [reply src/Driver/driver_configs.jl:276]: if we apply a cutoff filter that is the same degree or higher, it won't do anything (so is safe to apply the filter in the other direction).

Comment (valeriabarra) [reply src/Driver/driver_configs.jl:276]: I don't think so. Since currently there is an `assert` in the `Filter.jl` module. I tested this and got the following stack:

```
ERROR: LoadError: AssertionError: all(0 .<= Nc .<= N)
Stacktrace:
 [1] CutoffFilter(::DiscontinuousSpectralElementGrid{Float64,3,(5, 5, 5),216,Array,Tuple{Array{Float64,1},Array{Float64,1},Array{Float64,1}},Tuple{Array{Float64,2},Array{Float64,2},Array{Float64,2}},Array{Float64,3},Array{Float64,4},Array{Int64,1},Array{Int64,2},Array{Int64,3},ClimateMachine.Mesh.Topologies.StackedCubedSphereTopology{Float64,2},Tuple{Array{Float64,2},Array{Float64,2},Array{Float64,2}}}, ::Tuple{Int64,Int64,Int64}) at /home/valeria/CliMA/ClimateMachine.jl/src/Numerics/Mesh/Filters.jl:264
```

File: src/Numerics/DGMethods/DGModel.jl
Comment (jkozdon) [inline src/Numerics/DGMethods/DGModel.jl:555]:
```suggestion
            comp_stream = Filters.apply_async!(
                dg.state_gradient_flux,
                1:num_state_gradient_flux,
                dg.grid,
                dg.gradient_filter;
                dependencies = comp_stream
            )
```

File: src/Numerics/DGMethods/DGModel.jl
Comment (jkozdon) [inline src/Numerics/DGMethods/DGModel.jl:773]:
```suggestion
    if dg.tendency_filter !== nothing
        comp_stream = Filters.apply!(
            tendency,
            1:num_state_tendency,
            dg.grid,
            dg.tendency_filter;
            dependencies = comp_stream
        )
    end
    wait(device, comp_stream)
```

**Review by jkozdon — COMMENTED** [review]
Two questions:

- Hyperdiffusion should probably be filtered as well
- Shouldn't the filter be applied in 'init_ode_state' as well?

**Comment by jkozdon** [comment]
Looking good though!

**Comment by valeriabarra** [comment]
bors try

File: src/Numerics/DGMethods/DGModel.jl
Comment (valeriabarra) [inline src/Numerics/DGMethods/DGModel.jl:773]: I'm trying to debug this. 
I tested this with `experiments/TestCase/isothermal_zonal_flow.jl` and here `comp_stream = nothing` and the code breaks at the following line, with error:
```
ERROR: LoadError: MethodError: no method matching wait(::KernelAbstractions.CPU, ::Nothing)
```
I'm trying to understand why in this case the `apply_async!` returns nothing to the `Filters.apply!`. @simonbyrne , @jkozdon any clue?

Comment (valeriabarra) [reply src/Numerics/DGMethods/DGModel.jl:773]: Ah! I rolled back to commit [ac05980](https://github.com/CliMA/ClimateMachine.jl/pull/1953/commits/ac059805ed895efb7e08bf2e7e453a6f1c1094d1), that was before applying @jkozdon 's suggestion and it works. Hence, something needs to be improved, building on that suggestion. Thanks :)

File: src/Numerics/DGMethods/DGModel.jl
Comment (mwarusz) [inline src/Numerics/DGMethods/DGModel.jl:425]:
```suggestion
        comp_stream = Filters.apply_async!(
```

File: src/Numerics/DGMethods/DGModel.jl
Comment (mwarusz) [inline src/Numerics/DGMethods/DGModel.jl:766]:
```suggestion
        comp_stream = Filters.apply_async!(
```

Comment (valeriabarra) [reply src/Numerics/DGMethods/DGModel.jl:766]: Thanks @mwarusz but that still doesn't work :/

File: src/Driver/driver_configs.jl
Comment (valeriabarra) [reply src/Driver/driver_configs.jl:276]: @simonbyrne  I fixed this in my branch `valeria/fix-cutoff-degree-cl-option`. You can merge it here when the issue regarding the `apply_async!` is sorted out.

File: src/Numerics/DGMethods/DGModel.jl
Comment (mwarusz) [reply src/Numerics/DGMethods/DGModel.jl]: Weird, works for me. Just to make sure: I had two separate suggestions, your latest has the first one but not this one. Is this what you wanted ?

Comment (valeriabarra) [reply src/Numerics/DGMethods/DGModel.jl]: Oh, yes! Sorry! I had accidentally missed one of the two. Thanks! Just testing it now.

**Comment by valeriabarra** [comment]
bors try

File: src/Numerics/DGMethods/DGModel.jl
Comment (valeriabarra) [reply src/Numerics/DGMethods/DGModel.jl]: Update: This has been resolved by @mwarusz 's fixes! Thanks @jkozdon  and @mwarusz !!!

File: src/Driver/driver_configs.jl
Comment (valeriabarra) [reply src/Driver/driver_configs.jl]: Update: I pushed it to this branch. So it is fixed now and I dismissed my request for changes in this PR

**Comment by valeriabarra** [comment]
bors try-

**Comment by valeriabarra** [comment]
bors try

**Review by valeriabarra — APPROVED** [review]

**Comment by valeriabarra** [comment]
@simonbyrne I think this PR is ready to go. I will reduce the number of commits to a small number of significant commits to preserve authorship of all contributors to this PR, and then `r+` if that's ok. Thanks

**Comment by jkozdon** [comment]
Should we allow in some way for different filters in the `vertical` and `horiztonal`? Not sure if mixing polynomial orders is still in the cards, but if you wanted overintegration with this we might need two different filters

**Comment by valeriabarra** [comment]
That's what I thought too, when testing this and was wondering if we wanted to allow the filtering to happen only in one direction and for instance `nothing` in the other. We would need to erase all the `@assert (0 .<= Nc .<= N)` in the grid, if we want to go that way.

**Comment by simonbyrne** [comment]
Thanks everyone for the contributions, I really appreciate the help.

> Should we allow in some way for different filters in the `vertical` and `horiztonal`? Not sure if mixing polynomial orders is still in the cards, but if you wanted overintegration with this we might need two different filters

Can't you specify different orders for the cutoff filter? Or you mean apply different filters?

> That's what I thought too, when testing this and was wondering if we wanted to allow the filtering to happen only in one direction and for instance `nothing` in the other. We would need to erase all the `@assert (0 .<= Nc .<= N)` in the grid, if we want to go that way.

If `Nc == N`, won't that do nothing (since you're filtering to the same degree)?

**Comment by jkozdon** [comment]
> Can't you specify different orders for the cutoff filter? Or you mean apply different filters?

You're right, the cutoff filter takes a tuple of polynomial orders. I forgot!

> If `Nc == N`, won't that do nothing (since you're filtering to the same degree)?

'Nc' is the first cutoff mode. So you would need 'Nc = N + 1' would be a non-op (though the assert doesn't allow this). Would should probably allow for it in someway, but maybe that's a separate pr?

**Comment by valeriabarra** [comment]
> 'Nc' is the first cutoff mode. So you would need 'Nc = N + 1' would be a non-op (though the assert doesn't allow this). Would should probably allow for it in someway, but maybe that's a separate pr?

I would leave that to a separate PR

**Comment by simonbyrne** [comment]
I updated the merge. Any final comments @valeriabarra?

**Comment by valeriabarra** [comment]
Thanks Simon. No more comments from me. If we want to separate the directions and remove all the asserts `@assert (0 .<= Nc .<= N)`, I think it could belong to a separate PR.

**Pull request merged**
